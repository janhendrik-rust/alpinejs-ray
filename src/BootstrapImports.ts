/*
import { Ray, ray } from 'node-ray/web';
import Alpine from 'alpinejs';
import AlpineRayPlugin from 'alpinejs-ray';

window.ray = ray;
window.Ray = Ray;
window.axios = require('axios');
window.axios.defaults.headers.common['X-Requested-With'] = 'XMLHttpRequest';

window.Alpine = Alpine;
window.AlpineRayPlugin = AlpineRayPlugin;
window.AlpineRayPlugin.init();
window.AlpineRayPlugin.start();
*/

import { Ray, ray } from 'node-ray/web';
import axios from 'axios';
import AlpineRayPlugin from './index';
import { getWindow } from './lib/utils';
import { bootstrap } from './lib/bootstrap';

export const bootstrapImports = (Alpine: any, window: any = null) => {
    window = window ?? getWindow();

    return bootstrap({ Alpine, Ray, ray, axios, AlpineRayPlugin }, window);
};

export default bootstrapImports;
