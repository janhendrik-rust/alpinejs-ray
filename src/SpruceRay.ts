/* eslint-disable no-undef */

export class SpruceRay {
    public spruceInstance: any;
    public rayInstance: any;
    public trackRays: Record<string, any> = {};

    public constructor(rayFunc: any = null, spruceInstance: any = null) {
        this.rayInstance = rayFunc ?? globalThis.ray;
        this.spruceInstance = spruceInstance ?? globalThis.Spruce;
    }

    public static create(rayInstance: any = null, spruceInstance: any = null): SpruceRay {
        return new SpruceRay(rayInstance, spruceInstance);
    }

    protected init() {
        this.rayInstance = this.rayInstance ?? globalThis.ray() ?? null;
        this.spruceInstance = this.spruceInstance ?? globalThis.Spruce ?? null;
    }

    public hasSpruceInstallation() {
        return this.spruceInstance !== null;
    }

    protected isValidWatchName(name: string): boolean {
        return name.length > 0 && name.includes('.');
    }

    public watch(name: string) {
        if (!this.hasSpruceInstallation()) {
            return;
        }

        if (!this.isValidWatchName(name)) {
            return;
        }

        this.init();

        if (typeof this.trackRays[name] === 'undefined') {
            this.trackRays[name] = this.rayInstance;
        }

        this.spruceInstance.watch(name, value => {
            if (typeof this.trackRays[name] !== 'undefined') {
                const title = 'Spruce Watched Property Change';
                this.trackRays[name].table({ title, name, value });
            }
        });
    }

    public unwatch(name: string) {
        if (!this.hasSpruceInstallation()) {
            return;
        }

        if (!this.isValidWatchName(name)) {
            return;
        }

        this.init();

        if (typeof this.trackRays[name] !== 'undefined') {
            delete this.trackRays[name];
        }
    }
}

export default SpruceRay;
