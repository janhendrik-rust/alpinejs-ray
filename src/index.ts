//import { checkForRay } from './lib/utils';
import AlpineRayPlugin from './AlpineRayMagicMethod';

const alpineMethods = [AlpineRayPlugin];

const initializeLibrary = () => {
    alpineMethods.forEach(am => {
        am.init();
    });
};

initializeLibrary();

export default AlpineRayPlugin;
