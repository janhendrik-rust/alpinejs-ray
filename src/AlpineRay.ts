/* eslint-disable no-undef */

import { Ray } from 'node-ray/web';
import { SpruceRay } from './SpruceRay';

export class AlpineRay extends Ray {
    public static $version = '__BUILD_VERSION__';

    public spruceRay: SpruceRay | null = null;

    public spruce(): SpruceRay {
        if (this.spruceRay === null) {
            this.spruceRay = SpruceRay.create();
        }

        return <SpruceRay>this.spruceRay;
    }
}

export const ray = (...args: any[]) => {
    // @ts-ignore
    return AlpineRay.create().send(...args);
};

globalThis.ray = function (...args: any[]) {
    return ray(...args);
};

globalThis.AlpineRay = Ray;
