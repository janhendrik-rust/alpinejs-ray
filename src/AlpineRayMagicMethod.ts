/* eslint-disable no-undef */

import { getWindow, checkForAlpine, checkForAxios, highlightHtmlMarkup, filterObjectKeys } from './lib/utils';
import { ray } from './AlpineRay';
import { getAlpineRayConfig, AlpineRayConfig } from './AlpineRayConfig';

const AlpineRayMagicMethod = {
    initOnDocumentReady(window: any = null) {
        window = window ?? getWindow();

        window.document.addEventListener('readystatechange', () => {
            if (window.document.readyState === 'ready') {
                window.deferLoadingAlpine();
            }
        });
    },

    initDeferLoadingAlpine(window: any = null, rayInstance: any = null) {
        window = window ?? getWindow();
        rayInstance = rayInstance ?? ray;

        const config = getAlpineRayConfig(window);
        const alpine = window.deferLoadingAlpine || ((alpine: any) => alpine());

        window.deferLoadingAlpine = function (callback: CallableFunction) {
            AlpineRayMagicMethod.initOnComponentInitialized(config, window, rayInstance);
            AlpineRayMagicMethod.start(window, rayInstance);

            alpine(callback);
        };
    },

    initOnComponentInitialized(config: AlpineRayConfig | Record<string, unknown>, window: any = null, rayInstance: any = null) {
        window.Alpine.onComponentInitialized(el => {
            if (config.logComponentsInit ?? false) {
                rayInstance.table(
                    {
                        'component event': 'component init',
                        data: filterObjectKeys(el.$data, ['$el', '$watch', '$refs', '$nextTick']),
                        HTML: `<code class="text-sm text-black">${highlightHtmlMarkup(el.$el.outerHTML)}</code>`,
                    },
                    'alpine.js'
                );
            }

            if (config.logCustomEvents ?? false) {
                const eventRegex = new RegExp(/(?:@|x-on:)([\w-_.]+)=/g);
                const html: string = el.$el.outerHTML;
                const matches = html.matchAll(eventRegex);
                const eventNames: string[] = [];

                for (const match of matches) {
                    eventNames.push(match[1]);
                }

                eventNames.forEach(name => {
                    const nameParts = name.split('.');
                    const eventName: string = nameParts[0] ?? name;
                    const lastNamePart = nameParts[nameParts.length - 1];

                    window.addEventListener(eventName, e => {
                        if (eventName.includes('-') || (nameParts.length === 2 && lastNamePart === 'window')) {
                            rayInstance.table({
                                event: name,
                                payload: e.detail ?? null,
                            });
                        }
                    });
                });
            }
        });
    },

    init(window: any = null, rayInstance: any = null) {
        window = window ?? getWindow();
        rayInstance = rayInstance ?? ray;

        AlpineRayMagicMethod.initOnDocumentReady(window);
        AlpineRayMagicMethod.initDeferLoadingAlpine(window, rayInstance);
    },

    start(window: any = null, rayInstance: any = null) {
        window = window ?? getWindow();
        rayInstance = rayInstance ?? ray;

        checkForAxios(window);
        checkForAlpine(window);

        window.Alpine.addMagicProperty('ray', () => rayInstance);
    },
};

export default AlpineRayMagicMethod;
