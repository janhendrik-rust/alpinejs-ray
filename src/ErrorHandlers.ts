import { encodeHtmlEntities, filterObjectKeys, findParentComponent, getWindow, highlightHtmlMarkup } from './lib/utils';
import { ray } from './AlpineRay';

export function addErrorEventHandlers(window: any = null, rayInstance: any = null) {
    // @ts-ignore
    rayInstance = rayInstance ?? ray();
    window = window ?? getWindow();

    window.addEventListener('error', (errorEvent: any) => {
        if (errorEvent.error) {
            const { el, expression } = errorEvent.error;
            const parentComponent = findParentComponent(el);

            // component and parent components are not alpine components, so do nothing
            if (!parentComponent) {
                return;
            }

            // highlight the expression that triggered the error
            const parentComponentOuterHTML = highlightHtmlMarkup(parentComponent.outerHTML).replace(
                encodeHtmlEntities(expression),
                `<span class="text-red-700 bg-red-300 p-1">${encodeHtmlEntities(expression)}</span>`
            );

            const componentData = parentComponent.__x ?? { $data: {} };

            rayInstance.table(
                {
                    errorType: `alpine.js error`,
                    errorReason: errorEvent.error.toString(),
                    expression: `<code>${expression}</code>`,
                    srcElement: el.tagName.toLowerCase(),
                    componentHtml: `<code>${parentComponentOuterHTML}</code>`,
                    componentData: filterObjectKeys(componentData.$data, ['$el', '$watch', '$refs', '$nextTick']),
                },
                'ERROR'
            );
        }
    });

    window.addEventListener('unhandledrejection', (rejectionEvent: any) => {
        if (rejectionEvent.reason) {
            const { el, expression } = rejectionEvent.reason;
            const parentComponent = findParentComponent(el);

            // component and parent components are not alpine components, so do nothing
            if (!parentComponent) {
                return;
            }

            // highlight the expression that triggered the error
            const parentComponentOuterHTML = highlightHtmlMarkup(parentComponent.outerHTML).replace(
                encodeHtmlEntities(expression),
                `<span class="text-red-700 bg-red-300 p-1">${encodeHtmlEntities(expression)}</span>`
            );

            const componentData = parentComponent.__x ?? { $data: {} };

            rayInstance.table(
                {
                    errorType: `alpine.js ${rejectionEvent.type}`,
                    errorReason: rejectionEvent.reason.toString(),
                    expression: `<code>${expression}</code>`,
                    srcElement: el.tagName.toLowerCase(),
                    componentHtml: `<code>${parentComponentOuterHTML}</code>`,
                    componentData: filterObjectKeys(componentData.$data, ['$el', '$watch', '$refs', '$nextTick']),
                },
                'ERROR'
            );
        }
    });
}
