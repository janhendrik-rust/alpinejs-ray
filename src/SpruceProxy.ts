/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */

import { getWindow } from './lib/utils';
import { ray } from './AlpineRay';

const trackRays = {};

export class SpruceProxy {
    public spruceInstance: any = null;
    public window: any;
    public rayInstance: any;

    constructor(window: any = null, rayInstance: any = null) {
        this.window = window ?? getWindow();
        this.rayInstance = rayInstance ?? ray();
        this.spruceInstance = this.window.Spruce;
    }

    public checkForSpruce() {
        if (!this.hasSpruceInstallation()) {
            throw new Error('[alpinejs-ray] Spruce must be installed to use the SpruceProxy.');
        }
    }

    public hasSpruceInstallation() {
        return typeof this.window['Spruce'] !== 'undefined';
    }

    public spruce() {
        return this.spruceInstance;
    }

    public init() {
        if (this.hasSpruceInstallation()) {
            this.checkForSpruce();

            this.spruceInstance = this.window.Spruce;
            this.window.SpruceProxy = this;

            this.interceptStoreCreation();
        }
    }

    public interceptStoreCreation() {
        const displayTracking = this.displayTracking;

        const proxyHandler: ProxyHandler<any> = {
            get(target, key) {
                if (typeof target[key] === 'undefined' || !(<string>key ?? '').length) {
                    return;
                }

                displayTracking(<string>key, 'read/write', target[<string>key]);

                return target[key];
            },
            set(target, key, value) {
                if (!(<string>key ?? '').length) {
                    return false;
                }

                target[key] = value;

                displayTracking(<string>key, 'created', value);

                return true;
            },
        };

        this.window.Spruce.stores = new Proxy(this.window.Spruce.stores, proxyHandler);
    }

    public displayTracking(key: string, action: string, value: any) {
        if (typeof trackRays[<string>key] === 'undefined') {
            trackRays[<string>key] = this?.rayInstance() ?? ray();
        }

        const data = {
            title: `Spruce Store`,
            name: key,
            action: action,
            updated: new Date().getTime(),
            target: value,
        };

        trackRays[key] = trackRays[key].table(data);
    }

    public dumpStore(name: string) {
        const rayInstance = this?.rayInstance ?? ray;
        rayInstance().table(this.spruceInstance.stores[name]);
    }

    public dumpAllStores() {
        const rayInstance = this?.rayInstance ?? ray;
        rayInstance().table(this.spruceInstance.stores);
    }
}
