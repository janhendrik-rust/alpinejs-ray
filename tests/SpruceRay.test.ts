/* eslint-disable no-undef */

import { SpruceRay } from '../src/SpruceRay';

let spruceRayInstance: SpruceRay,
    fakeSpruce: FakeSpruce,
    fakeRay: FakeRay,
    testState = { rayPayloadHistory: [] };

class FakeRay {
    table(...args: any[]): FakeRay {
        return this.send('table', args);
    }

    send(type: string, ...args: any): FakeRay {
        // @ts-ignore
        testState.rayPayloadHistory.push({ type, args });
        return this;
    }
}

class FakeSpruce {
    public stores: Record<string, any> = {
        mydata: {
            one: 1,
        },
    };

    public watches: Record<string, CallableFunction> = {};

    watch(name: string, callback: CallableFunction) {
        this.watches[name] = callback;
    }

    changeStoreProp(store: string, prop: string, newValue: any) {
        this.stores[store][prop] = newValue;

        if (typeof this.watches[`${store}.${prop}`] !== 'undefined') {
            this.watches[`${store}.${prop}`](newValue);
        }
    }

    store(name: string) {
        if (typeof this.stores[name] === 'undefined') {
            this.stores[name] = {};
        }
        return this.stores[name];
    }
}

const fakeRayFunc = (...args: any[]) => {
    return new FakeRay().send('log', ...args);
};

beforeEach(() => {
    testState = {
        rayPayloadHistory: [],
    };

    fakeRay = new FakeRay();

    fakeSpruce = new FakeSpruce();

    spruceRayInstance = SpruceRay.create(fakeRay, fakeSpruce);
});

it('can watch a store property', () => {
    spruceRayInstance.watch('mydata.one');

    expect(fakeSpruce.watches['mydata.one']).not.toBeUndefined();
});

it('sends an update to Ray when a watched property changes', () => {
    spruceRayInstance.watch('mydata.one');
    fakeSpruce.changeStoreProp('mydata', 'one', 'new_value');

    expect(testState.rayPayloadHistory).toMatchSnapshot();
});

it('unwatches a watched store property', () => {
    spruceRayInstance.watch('mydata.one');

    expect(spruceRayInstance.trackRays['mydata.one']).not.toBeUndefined();

    spruceRayInstance.unwatch('mydata.one');

    expect(spruceRayInstance.trackRays['mydata.one']).toBeUndefined();
});
