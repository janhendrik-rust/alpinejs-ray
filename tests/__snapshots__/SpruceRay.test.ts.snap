// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sends an update to Ray when a watched property changes 1`] = `
Array [
  Object {
    "args": Array [
      Array [
        Object {
          "name": "mydata.one",
          "title": "Spruce Watched Property Change",
          "value": "new_value",
        },
      ],
    ],
    "type": "table",
  },
]
`;
